engine/access/ingestion2/results_forest.go

```diff
@@ -5,35 +5,179 @@ import (
+//     may respond with a dedicated sentinel informing the caller that it already has pruned up to view 70.
+//     Nevertheless, from the perspective of the ResultsForest, it should not be treated as a critical exception,
+//     because the information that all blocks up to view 60 are allowed to be pruned is a subset of the information
+//     the forest got before when being told that all blocks up to view 70 can be pruned.
+//   - Repeated addition of the same result and/or receipt should be a no-op.
+//   - <additional conditions?>
+//
+// The ResultsForest mempool supports pruning by view:
 // only results descending from the latest sealed and finalized result are relevant.
 // By convention, the ResultsForest always contains the latest sealed result. Thereby, the
 // ResultsForest is able to determine whether results for a block still need to be processed or
 // can be orphaned (processing abandoned). Hence, we prune all results for blocks _below_ the
-// latest block with a finalized seal.
+// latest block with a finalized seal, once it has been persisted in the database.
 // All results for views at or above the pruning threshold are retained, explicitly including results
 // from execution forks or orphaned blocks even if they conflict with the finalized seal. However, such
 // orphaned forks will eventually stop growing, because either (i) a conflicting fork of blocks is
 // finalized, which means that the orphaned forks can no longer be extended by new blocks or (ii) a
 // rogue execution node pursuing its own execution fork will eventually be slashed and can no longer
 // submit new results.
-// Nevertheless, to utilize resources efficiently, the ResultsForest tried to avoid processing execution
+// Nevertheless, to utilize resources efficiently, the ResultsForest tries to avoid processing execution
 // forks that conflict with the finalized seal.
 //
+// Pruning and abandoning processing of results:
+//   - For liveness, the ResultsForest must guarantee that all sealed results it knows about are
+//     eventually marked as processable.
+//   - For results that are not yet sealed, only results descending from the latest sealed result 𝓼 should
+//     ideally be processed. The ResultsForest is allowed to process results optimistically (resources
+//     permitting), accepting the possibility that some results will turn out to be orphaned later.
+//   - To utilize resources efficiently, the ResultsForest tries to avoid processing execution forks that
+//     conflict with the finalized seal. Furthermore, it attempts to cancel already ongoing processing once
+//     it concludes that a fork was abandoned. Specifically:
+//     (a) processing of execution forks that conflict with known sealed results are be abandoned (implemented)
+//     (b) processing of execution forks whose executed blocks conflict with finalized blocks can be abandoned
+//     (not yet implemented)
```

Comment:
This actual is implemented here:
https://github.com/onflow/flow-go/pull/7523/files#diff-4240b04727c486c182038c6eb280cbf969cbf3412ea5cbc58877407ec19648f5R361-R381